apps/web/src/connectors/duo-redirect.ts

```
@@ -5,24 +5,63 @@ require("./duo-redirect.scss");
 const mobileDesktopCallback = "bitwarden://duo-callback";
 
 window.addEventListener("load", () => {
+  const redirectUrl = getQsParam("duoFramelessUrl");
+  const handOffMessage = getQsParam("handOffMessage");
+
+  if (redirectUrl) {
+    redirectToDuoFrameless(redirectUrl, handOffMessage);
+    return;
+  }
+
   const client = getQsParam("client");
   const code = getQsParam("code");
+  const state = getQsParam("state");
 
   if (client === "web") {
     const channel = new BroadcastChannel("duoResult");
 
-    channel.postMessage({ code: code });
+    channel.postMessage({ code: code, state: state });
     channel.close();
 
     processAndDisplayHandoffMessage();
   } else if (client === "browser") {
-    window.postMessage({ command: "duoResult", code: code }, "*");
+    window.postMessage({ command: "duoResult", code: code, state: state }, "*");
     processAndDisplayHandoffMessage();
   } else if (client === "mobile" || client === "desktop") {
-    document.location.replace(mobileDesktopCallback + "?code=" + encodeURIComponent(code));
+    processAndDisplayHandoffMessage();
+    document.location.replace(
+      mobileDesktopCallback +
+        "?code=" +
+        encodeURIComponent(code) +
+        "&state=" +
+        encodeURIComponent(state),
+    );
   }
 });
 
+/**
+ * In order to set a cookie with the hand off message, some clients need to use
+ * this connector as a middleman to set the cookie before continuing to the duo url
+ * @param redirectUrl the duo auth url
+ * @param handOffMessage message to save as cookie
+ */
+function redirectToDuoFrameless(redirectUrl: string, handOffMessage: string) {
+  let validateUrl;
+  try {
+    validateUrl = new URL(redirectUrl);
+  } catch (error) {
+    return;
+  }
+  let parts = validateUrl.hostname.split(".");
+  parts = parts.slice(parts.length - 2);
+  const hostname = parts.join(".");
+  if (validateUrl.protocol !== "https:" || hostname !== "duosecurity.com") {
+    return;
+  }
```

Two questions:
(1) Are we worried about duo subdomains like `eu.duosecurity.com`?  Is that a possibility in our implementation? We could make the check more flexible with something like `!hostname.endsWith("duosecurity.com")`?
(2) Should we throw an error along with the return to make debugging potential issues easier?

---

Good calls, fixed.